draft-gpew-priv-ppm.md

```diff
@@ -629,268 +653,638 @@ encoded value of the following form:
+phase, then `out` is the helper's output share, in which case the helper
+finishes. Otherwise, the helper continues with `out` as its next VDAF message.
+
+### Aggregating Output Shares {#out-to-agg-share}
+
+Once processing of a report share reaches the FINISHED state, the aggregator
+stores the the recovered output share until the batch to which it pertains is
+collected. To aggregate the output shares, the aggregator runs the aggregation
+algorithm specified by the VDAF:
+
+~~~
+agg_share = output_to_aggregate_shares(agg_param,
+                                       output_shares)
+~~~
+
+Note that for most VDAFs, it is possible to aggregate output shares as they
+arrive rather than wait until the batch is collected. To do so however, it is
+necessary to enforce the batch parameters as described in
+{{batch-parameter-validation}} so that the aggregator knows which aggregate
+share to update.
+
+The leader needs to store a report for as long as it can be collected. Once a
+report has been collected `max_batch_lifetime` times, the leader SHOULD remove
+the report from storage.
+
+### Aggregate Messages {#aggregate-message-auth}
+
+The following messages are defined in this section and are used to aggregate a
+number of reports simultaneously:
+
+~~~
+opaque AggregationJobID[32];
+
+enum {
+  agg_init_req(0),
+  agg_req(1),
+} AggregateReqType;
 
 struct {
-  Nonce nonce;
-  opaque message<0..2^16-1>; // VDAF message
-} AggregateSubResp;
+    AggregateReqType msg_type;
+    select (msg_type) {
+        agg_init_req: AggregateInitReq;
+        agg_req: AggregateContinueReq;
+    }
+    opaque tag[32];
+} AggregateReq;
 ~~~
 
-The helper handles each sub-request `AggregateSubReq` as follows. It first looks
-up the HPKE config and corresponding secret key associated with
-`helper_share.config_id`. If not found, then the sub-response consists of an
-"unrecognized config" alert. [TODO: We'll want to be more precise about what
-this means. See issue#57.] Next, it attempts to decrypt the payload with the
-following procedure:
+As discussed in {aggregation-flow}, the leader may choose to shard the task of
+aggregating a large number of reports into any number of sub-batches. The
+process of aggregating the reports in a leader-defined sub-batch is called an
+*aggregation job*. An `AggregationJobID` is a globally unique sequence of bytes
+chosen by the leader to identify an aggregation job. It is RECOMMENDED that this
+be set to a random string output by a cryptographically secure pseudorandom
+number generator. The aggregation job ID is used by aggregators to look up
+preparation state for the reports included in the sub-batch.
+
+The last 32 bytes of each message is an HMAC-SHA256 tag computed over the
+serialized message excluding the `tag` field itself. The key that is used is the
+`agg_auth_key` shared by the aggregators and configured for the given task.
+
+Upon receiving an `AggregateReq` message, an aggregator MUST verify the tag before
```

Comment:
Indeed. We could also add a requirement that the size of the encoded values must be known (which is true for the top-layer messages that are being tagged, since they are being passed around as HTTP requests/responses); then the `tag` is just the last 32 bytes.

Reply:
The same pattern exists for TLS, I don't think this is a big problem. On the other hand I would advocate for being conservative about what we MAC. In particular I wouldn't want the disccriminator or length bits to not be covered by the MAC.